storage_controller/src/http.rs

```diff
@@ -87,19 +89,33 @@ fn get_state(request: &Request<Body>) -> &HttpState {
 }
 
 /// Pageserver calls into this on startup, to learn which tenants it should attach
-async fn handle_re_attach(mut req: Request<Body>) -> Result<Response<Body>, ApiError> {
+async fn handle_re_attach(req: Request<Body>) -> Result<Response<Body>, ApiError> {
     check_permissions(&req, Scope::GenerationsApi)?;
 
+    let mut req = match maybe_forward(req).await {
+        ForwardOutcome::Forwarded(res) => {
+            return res;
+        }
+        ForwardOutcome::NotForwarded(req) => req,
+    };
+
     let reattach_req = json_request::<ReAttachRequest>(&mut req).await?;
     let state = get_state(&req);
     json_response(StatusCode::OK, state.service.re_attach(reattach_req).await?)
 }
 
 /// Pageserver calls into this before doing deletions, to confirm that it still
 /// holds the latest generation for the tenants with deletions enqueued
-async fn handle_validate(mut req: Request<Body>) -> Result<Response<Body>, ApiError> {
+async fn handle_validate(req: Request<Body>) -> Result<Response<Body>, ApiError> {
     check_permissions(&req, Scope::GenerationsApi)?;
 
+    let mut req = match maybe_forward(req).await {
```

Comment:
Oof, this is really repetitive.

Since this logic only requires `Request`, couldn't it be implemented as a wrapper to all these request handlers?  i.e. all the places we currently just use `named_request_span` we could use some wrapper that does the maybe-forwarding.

Reply:
Yeah, me and Joonas tried to come up with a nicer abstraction, but didn't find a nice one at the time. I'll give it another go (probably in a follow-up).

Reply:
Will merge as is to see if this works well in staging and follow-up if it's the case.